apollo-federation/src/query_plan/optimize.rs

```diff
@@ -765,3 +825,410 @@ impl SelectionSet {
+        // - Decide in reverse dependency order, so that at each step, the fragment being visited
+        //   has following properties:
+        //   - It is either indirectly used by a previous fragment; Or, not used directly by any
+        //     one visited & retained before.
+        //   - Its usage count should be correctly calculated assuming dropped fragments were expanded.
+        // - Also, we can take advantage of the fact that `NamedFragments` is already sorted in
+        //   dependency order.
+        let min_usage_to_optimize: i32 = min_usage_to_optimize.try_into().unwrap_or(i32::MAX);
+        let original_size = self.size();
+        let mut indirect_deps: HashSet<Name> = HashSet::new();
+        self.retain_rev(|name, fragment| {
+            let usage_count = usages.get(name).copied().unwrap_or_default();
+            if usage_count >= min_usage_to_optimize || indirect_deps.contains(name) {
+                // Update `indirect_deps` with transitive dependencies.
+                for name in fragment.fragment_usages().keys() {
+                    indirect_deps.insert(name.clone());
+                }
+                true // Keep this
+            } else {
+                // Estimate the new usage count after expanding the `fragment`.
+                Self::update_usages(&mut usages, fragment, usage_count);
+                false // Drop this
+            }
+        });
+
+        // Short-circuiting: Nothing was dropped (fully used) => Nothing to change.
+        if self.size() == original_size {
+            return Ok(selection_set.clone());
+        }
+
+        selection_set.retain_fragments(self)
+    }
+
+    fn update_usages(usages: &mut HashMap<Name, i32>, fragment: &Node<Fragment>, usage_count: i32) {
+        let mut inner_usages = HashMap::new();
+        fragment.collect_used_fragment_names(&mut inner_usages);
+
+        for (name, inner_count) in inner_usages {
+            *usages.entry(name).or_insert(0) += inner_count * usage_count;
+        }
+    }
+}
+
+//=============================================================================
+// `optimize` methods (putting everything together)
+
+impl Selection {
+    fn optimize(
+        &self,
+        fragments: &NamedFragments,
+        validator: &mut FieldsConflictMultiBranchValidator,
+    ) -> Result<Selection, FederationError> {
+        match self {
+            Selection::Field(field) => field.optimize(fragments, validator),
+            Selection::FragmentSpread(_) => Ok(self.clone()), // Do nothing
+            Selection::InlineFragment(inline_fragment) => {
+                inline_fragment.optimize(fragments, validator)
+            }
+        }
+    }
+}
+
+impl FieldSelection {
+    fn optimize(
+        &self,
+        fragments: &NamedFragments,
+        validator: &mut FieldsConflictMultiBranchValidator,
+    ) -> Result<Selection, FederationError> {
```

Comment:
Do you think it'll be easier to reason about if this returns a `FieldSelection` instead? Something like this:

```suggestion
    fn optimize(
        mut self,
        fragments: &NamedFragments,
        validator: &mut FieldsConflictMultiBranchValidator,
    ) -> Result<Self, FederationError> {
```

I am thinking we are walking down the tree from the `SelectionSet` root, modifying and optimising it's potential children. It might be easier if the results bubble up as their own respective types, and we handle converting them into a `Selection` at one of the top level nodes?

Reply:
Potentially might not work due to borrow checker. But I guess what I am trying to say is that I am finding it difficult to follow that `FieldSelection`/`InlineFragmentSelection`/etc all optimize, theoretically their own values, but they don't actually always modify their own selection_set field _and also_ don't return the `selection_set` but a `selection`

Reply:
Indeed. In the JS version, `optimze()` function was a virtual function with the same return type (`Selection`). But, in Rust, we don't have do that. The wrinkle here is that `InlineFragmentSelection` can be optimized into `FragmentSpreadSelection`. So, it makes sense to add a new enum `InlineOrFragmentSelection`, which helps to eliminate one `unreachable!()` branch case.